core/src/main/scala/akka/persistence/r2dbc/journal/JournalDao.scala

```diff
@@ -242,46 +242,58 @@ private[r2dbc] class JournalDao(journalSettings: R2dbcSettings, connectionFactor
     result
   }
 
-  def deleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
-    val entityType = PersistenceId.extractEntityType(persistenceId)
-    val slice = persistenceExt.sliceForPersistenceId(persistenceId)
-
-    val deleteMarkerSeqNrFut =
-      if (toSequenceNr == Long.MaxValue)
-        readHighestSequenceNr(persistenceId, 0L)
-      else
-        Future.successful(toSequenceNr)
-
-    deleteMarkerSeqNrFut.flatMap { deleteMarkerSeqNr =>
-      def bindDeleteMarker(stmt: Statement): Statement = {
-        stmt
-          .bind(0, slice)
-          .bind(1, entityType)
-          .bind(2, persistenceId)
-          .bind(3, deleteMarkerSeqNr)
-          .bind(4, "")
-          .bind(5, "")
-          .bind(6, 0)
-          .bind(7, "")
-          .bind(8, Array.emptyByteArray)
-          .bind(9, true)
-      }
-
-      val result = r2dbcExecutor.update(s"delete [$persistenceId]") { connection =>
-        Vector(
+  def deleteEventsTo(persistenceId: String, toSequenceNr: Long, neverUsePersistenceIdAgain: Boolean): Future[Unit] = {
+    val result: Future[Long] =
+      if (neverUsePersistenceIdAgain) {
+        r2dbcExecutor.updateOne(s"delete [$persistenceId]") { connection =>
           connection
             .createStatement(deleteEventsSql)
             .bind(0, persistenceId)
-            .bind(1, toSequenceNr),
-          bindDeleteMarker(connection.createStatement(insertDeleteMarkerSql)))
+            .bind(1, toSequenceNr)
+        }
+      } else {
+
+        val entityType = PersistenceId.extractEntityType(persistenceId)
+        val slice = persistenceExt.sliceForPersistenceId(persistenceId)
+
+        val deleteMarkerSeqNrFut =
+          if (toSequenceNr == Long.MaxValue)
+            readHighestSequenceNr(persistenceId, 0L)
+          else
+            Future.successful(toSequenceNr)
+
+        deleteMarkerSeqNrFut.flatMap { deleteMarkerSeqNr =>
+          def bindDeleteMarker(stmt: Statement): Statement = {
+            stmt
+              .bind(0, slice)
+              .bind(1, entityType)
+              .bind(2, persistenceId)
+              .bind(3, deleteMarkerSeqNr)
+              .bind(4, "")
+              .bind(5, "")
+              .bind(6, 0)
+              .bind(7, "")
+              .bind(8, Array.emptyByteArray)
+              .bind(9, true)
+          }
+
+          r2dbcExecutor
+            .update(s"delete [$persistenceId]") { connection =>
+              Vector(
+                connection
+                  .createStatement(deleteEventsSql)
+                  .bind(0, persistenceId)
+                  .bind(1, toSequenceNr),
```

Comment:
Maybe I'm missing sth, but for very long journals for a given persistenceId, deleting millions of rows in one transaction will be very problematic. Big deletes are pretty slow and create a lot of contention for other statements.

Reply:
If we would add support for it here we would lookup lowest and highest sequence number and then split that up in multiple transactions executed sequentially in seq number order. Do you have other suggestion? Maybe not that difficult to add? Would you like to take a stab at that in a separate PR?

Reply:
Yeah, that was my idea as well. To run it in some recursive/iterate way. I could give it a try:)